### PR TITLE
Updated oneCCL to use the new API

### PR DESCRIPTION
**Pull request by shirosankaku**
Signed-off-by: Yana Shchyokotova <yana.shchyokotova@intel.com>

Updated oneCCL related files with oneCCL new API. No need to update docs and write extra tests.

oneCCL v1.0 specification
https://oneapi-src.github.io/oneAPI-spec/elements/oneCCL/source/index.html

Upd: oneCCL repo in Dockerfile.test.cpu will be changed soon to the official one.

### CODE REVIEW

**Comment by shirosankaku** [comment]
Hello, @tgaddair !
Could you, please, review ? 
Also, we faced some issues with Gloo test when it fails due to test_colocator assert. Could you give some piece of advice regarding handling this?

**Comment by tgaddair** [comment]
Hey @shirosankaku, that test should be fixed on master.  Can you try rebasing and pushing?

File: Dockerfile.test.cpu [hunk 1]
```diff
@@ -74,21 +75,21 @@ RUN if [[ ${MPI_KIND} == "OpenMPI" ]]; then \
             cd /usr/local && tar -zxf /tmp/openmpi-3.0.0-bin.tar.gz && ldconfig && \
             echo "mpirun -allow-run-as-root -np 2 -H localhost:2 -bind-to none -map-by slot -mca mpi_abort_print_stack 1" > /mpirun_command; \
     elif [[ ${MPI_KIND} == "ONECCL" ]]; then \
-        wget --progress=dot:mega -O /tmp/oneccl.tar.gz https://github.com/oneapi-src/oneCCL/archive/master.tar.gz && \
+        wget --progress=dot:mega -O /tmp/oneccl.tar.gz https://github.com/shirosankaku/oneCCL/archive/${CCL_PACKAGE}.tar.gz && \
```
Comment (tgaddair) [inline Dockerfile.test.cpu:78]: Is this the right org for this repo?  Should it be `oneapi-src`?

File: horovod/common/ops/ccl_operations.cc [hunk 2]
```diff
@@ -17,255 +17,436 @@
...
+        public:
+            // Return stream/communicator (by const reference) for given TensorTableEntry.
+            // We keep one struct per device/host.
+            // We memoize them in a map. If for the given e.device we already have a struct we simply return it.
+            // Otherwise we initialize kvs if needed, create a new struct, store and return it.
+            // Before returning the looked-up/new struct we also set the curr_ pointer.
+            const ccl4hvd & GetCCL4HVD(const TensorTableEntry & e, const HorovodGlobalState* global_state) {
+                auto resit = this->contexts_.find(e.device);
+                
+                if(resit == this->contexts_.end()) {
```
Comment (tgaddair) [inline horovod/common/ops/ccl_operations.cc:71]: Please `clang-format` this file.

File: horovod/common/ops/ccl_operations.h [hunk 3]
```diff
@@ -21,72 +21,106 @@
...
+        class CCLOpContext;
+
+        // CCL Context used to control CCL as a whole from the outside
+        class CCLContext {
+        public:
+            CCLContext();
+            void Initialize();
+            void Finalize();
+            bool IsInited() { return opctxt_ != nullptr; }
+            CCLOpContext * opctxt_;
```
Comment (tgaddair) [inline horovod/common/ops/ccl_operations.h:44]: Please `clang-format` this file.

File: horovod/common/ops/collective_operations.cc [hunk 4]
```diff
@@ -142,9 +142,9 @@ Status AllgatherOp::AllocateOutput(std::vector<TensorTableEntry>& entries,
     for (int rc = 0; rc < global_size; ++rc) {
       auto component_size = tensor_sizes[ec * global_size + rc];
       total_entry_dimension_size += component_size;
-      recvcounts[rc] += component_size * single_slice_shape.num_elements();
-      entry_component_sizes[ec][rc] =
-          component_size * single_slice_shape.num_elements();
+      if(recvcounts) recvcounts[rc] += component_size * single_slice_shape.num_elements();
```
Comment (tgaddair) [inline horovod/common/ops/collective_operations.cc:145]: Please separate the if's onto separate lines and enclose within a scope:

```
if (recvcounts) {
    ...
}
```

Also, please note spacing with `if (`.

**Review by tgaddair — COMMENTED** [review]
Looks good.  Please do a quick formatting pass following the comments.

File: Dockerfile.test.cpu (hunk 1 above)
Comment (mshiryaev) [reply Dockerfile.test.cpu:78]: @tgaddair - we will replace fork URL by official repo URL right after oneCCL Gold release in few weeks
we will notify you
before that please review PR and keep it open

**Comment by tgaddair** [comment]
@shirosankaku @mshiryaev looks like oneCCP tests are failing the nightly tests.  Will landing this PR resolve the issues?

https://buildkite.com/horovod/horovod/builds/4394#41abd707-a002-4208-ad26-d1bc5c61821f

**Comment by shirosankaku** [comment]
Hello @tgaddair !
Could you give some advice regarding the issue with those packages 
Cannot install torch==1.2.0+cpu and torchvision==0.4.1+cpu because these package versions have conflicting dependencies
I Dockerfile.test.cpu it's clearly stated to use torchvision of version 0.4.0 but the script still tries to install 0.4.1

**Comment by tgaddair** [comment]
Hey @shirosankaku, looks like tests are passing now, is this ready to land?

**Review by tgaddair — APPROVED** [review]
LGTM.  Going to land this now to unblock, please see #2499 for follow-up following oneCCL gold release.

**Pull request merged**